### PR TITLE
ability to create + attach volume to new instance

### PR DESCRIPTION
**Pull request by amitofs**

### CODE REVIEW

File: gonzo/clouds/compute.py [hunk 1]
```diff
@@ -146,8 +147,47 @@ def create_instance(self, image_name, name, owner, user_data=None,
...
+
+        self.compute_session.attach_volume(
+            node=instance,
+            volume=created_volume,
+            device='/dev/xvdf'
+        )
+
+    def wait_until_volume_available(self, volume):
+        while not self.volume_is_created(volume):
+            print "Volume Being Created"
```
Comment (davidszotten) [inline gonzo/clouds/compute.py]: i would not expect a library method to be printing stuff

Comment (davidszotten) [inline gonzo/clouds/compute.py]: probably want some kind of `sleep` here

File: gonzo/clouds/compute.py [hunk 2]
```diff
@@ -146,8 +147,47 @@ def create_instance(self, image_name, name, owner, user_data=None,
...
+        return True
+
+    def volume_is_created(self, volume):
+        for vol in self.compute_session.list_volumes():
+            if vol.id == volume.id:
+                if vol.extra['state'] == 'available':
+                    return True
```
Comment (davidszotten) [inline gonzo/clouds/compute.py]: you probably want to distinguish "still pending" (return false) from "unknown volume" (raise exception?)

File: gonzo/clouds/compute.py [hunk 3]
```diff
@@ -146,8 +147,47 @@ def create_instance(self, image_name, name, owner, user_data=None,
...
+
+    def get_az_of_instance(self, instance):
+        instance_az = instance.extra['gonzo_az']
+
+        if len(self.list_availability_zones()) == 1:
```
Comment (davidszotten) [inline gonzo/clouds/compute.py]: how come we need a special case for this?

File: gonzo/clouds/compute.py [hunk 4]
```diff
@@ -146,8 +147,47 @@ def create_instance(self, image_name, name, owner, user_data=None,
...
+            return self.list_availability_zones()[0]
+
+        for az in self.list_availability_zones():
+            if az.name == instance_az:
+                return az
```
Comment (davidszotten) [inline gonzo/clouds/compute.py]: again, what happens if we don't find it?

File: tests/integration/test_end_to_end.py [hunk 5]
```diff
@@ -63,6 +63,13 @@ def openstack_session(fake_get_config):
     sec_groups = openstack.compute_session.ex_list_security_groups()
     for sec_group in sec_groups[1:]:  # Skip default group
         openstack.compute_session.ex_delete_security_group(sec_group)
+
+    print "deleting volumes"
```
Comment (davidszotten) [inline tests/integration/test_end_to_end.py]: print?

File: tests/integration/test_end_to_end.py [hunk 6]
```diff
@@ -99,6 +107,15 @@ def test_end_to_end(capsys, fake_dns, fake_get_config, openstack_session):
         instance_name[-2:]
     )
 
+    # check ebs created
+    time.sleep(10)
```
Comment (davidszotten) [inline tests/integration/test_end_to_end.py]: why the sleep? doesn't the launch method already wait?

File: tests/integration/test_end_to_end.py [hunk 7]
```diff
@@ -99,6 +107,15 @@ def test_end_to_end(capsys, fake_dns, fake_get_config, openstack_session):
...
+    volume = openstack_session.compute_session.list_volumes()[0]
```
Comment (davidszotten) [inline tests/integration/test_end_to_end.py]: this formulation is slightly unusual. if something is wrong, the assignment will raise `IndexError`, whereas your assert below is unlikely to ever fail, unless `list_volumes` returns e.g. `[False]` or `[None]`.

more common might be
`assert s.list_volumes()  # makes sure this isn't empty`

or even 
```
volumes = s.list_volumes())
assert len(volumes) == 1
```

and then just
`volume = volumes()[0]`

File: tests/integration/test_end_to_end.py (hunk 6 above)
Comment (amitofs) [inline tests/integration/test_end_to_end.py]: the launch method waits for the volume to be created. this sleep waits for the volume to be attached to the instance

File: gonzo/clouds/compute.py (hunk 2 above)
Comment (amitofs) [inline gonzo/clouds/compute.py]: done

Comment (davidszotten) [inline gonzo/clouds/compute.py]: almost. what happens if you find the volume but the state is neither "available" nor "creating"?

File: CHANGES [hunk 8]
```diff
@@ -1,3 +1,9 @@
+Version 0.4.1
```
Comment (davidszotten) [inline CHANGES]: for next time: 

for a whole new feature i might have gone for a larger bump, e.g. `0.5.0`.

**Pull request merged**